---
title: "Transcript clustering in multiple tissues"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

```{r set_type}
#set testing to FALSE to run full gene expression networks and to TRUE
#to run a subset.
#testing = TRUE
testing = FALSE
is.interactive = FALSE
#is.interactive = TRUE
#net.type = "signed"
net.type = "unsigned"
min.mod.size = 5
tissues = c("Islet", "Liver", "Adipose", "SkeletalMuscle")

args <- commandArgs(trailingOnly=T)
tissue.idx <- as.numeric(args[1])
clust.type <- args[2]

tissue.type <- tissues[tissue.idx]

if(is.na(tissue.type)){
  tissue.type = "SkeletalMuscle"
}

if(is.na(clust.type)){
  clust.type = "WGCNA"
}

library(here)
clust.dir <- here("Results", paste0(clust.type, "_Clusters"))
if(!file.exists(clust.dir)){dir.create(clust.dir)}
results.dir <- here("Results", paste0(clust.type, "_Clusters"), tissue.type)
if(!file.exists(results.dir)){dir.create(results.dir)}
```

## Gene modules in `r tissue.type` clustered by `r clust.type`

The purpose of this workflow is to cluster bulk tissue gene expression 
from the DO mice with WGCNA and CoExpNets. Downstream analyses will
compare the clustering from these two methods.

This workflow cluster transcripts separately in the specified tissue,
and runs Islet by default. Other possible tissues are Adipose, Liver,
and SkeletalMuscle.


```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "gprofiler2", "CoExpNets", "qtl2", "WGCNA")
load_libraries(all.packages)
```


```{r read_data}
exp.file <- here("Data", "RDS_datasets_tissues", paste0(tissue.type, ".RDS"))
tissue.exp <- readRDS(exp.file)
```

Also collect clinical traits.

```{r pheno}
pheno <- read.csv(here("Data", "DO_clinical_phenotypes.csv"), stringsAsFactors = FALSE, row.names = 1)
#subset to numeric traits
num.pheno <- apply(pheno[,11:30], 2, as.numeric)
rownames(num.pheno) <- rownames(pheno)
tissue.covar <- tissue.exp$covar.matrix
adj.pheno <- adjust(num.pheno, tissue.covar)
gene.table <- tissue.exp$annot.mrna
```

## Filter Expression
Filter the expression matrices to include only genes that have at least
a minimum amount of expression. 

Here we use the raw expression matrix to select transcripts from the rank Z
normalized expression matrix. 

```{r adjust_covar}
min.mean = 10
tissue.trans <- which(colMeans(tissue.exp$data$raw) > min.mean)
tissue.expr <- tissue.exp$data$rz[,tissue.trans]
adj.expr <- adjust(tissue.expr, tissue.covar)
```

## Cluster gene expression using CoExpNets or WGCNA

CoExpNets is WGCNA plus a k-means clustering step that redistributes 
genes across modules. It generally identifies clusters with more even 
size distribution and more specific enrichments than WGCNA.

```{r coExpNets}
if(clust.type == "CoExpNet"){
    #Clustering takes a long time
    #set fullAnnotation to F because function defaults to human.
    #we will do this ourselves by hand.
    net.file <- get.files(results.dir, want = c("net", "rds"), dont.want = "pdf", 
    full.names = TRUE, ignore.case = FALSE)
    if(length(net.file) == 0){
    net.file = CoExpNets::getDownstreamNetwork(tissue=tissue.type,
            net.type = net.type, debug=testing, expr.data = adj.expr,
            job.path= results.dir, save.plots = TRUE, fullAnnotation = F,
            min.cluster.size = min.mod.size)
    }
    clust.net <- readRDS(net.file)
}
```


```{r wgcna}
if(clust.type == "WGCNA"){
    wgcna.results <- here("Results", "WGCNA_Clusters")
    wgcna.net.file <- file.path(results.dir, paste0("net.RDS"))
    if(!file.exists(wgcna.net.file)){
    enableWGCNAThreads()

    if(testing){
      rnd.sample <- sample(1:ncol(adj.expr), 1500)
    }else{
      rnd.sample <- 1:ncol(adj.expr)
    }
    powers = c(c(1:10), seq(from = 12, to=20, by=2))
    sft = pickSoftThreshold(adj.expr[,rnd.sample], 
    powerVector = powers, verbose = 5)
    use.power <- sft$powerEstimate
    if(is.na(use.power)){use.power = 6}

    clust.net = blockwiseModules(adj.expr[,rnd.sample], 
    power = use.power, TOMType = net.type, 
    minModuleSize = min.mod.size, reassignThreshold = 0, 
    mergeCutHeight = 0.15, numericLabels = FALSE, 
    pamRespectsDendro = FALSE, saveTOMs = FALSE, deepSplit = 4)

    saveRDS(clust.net, wgcna.net.file)
    }else{
    clust.net <- readRDS(wgcna.net.file)
    }

    mergedColors = labels2colors(clust.net$colors)
    wgcna.modules <- clust.net$colors

    pdf(file.path(results.dir, "Modules.pdf"))
    for(i in 1:length(clust.net$dendrograms)){
        plotDendroAndColors(clust.net$dendrograms[[i]], 
    mergedColors[clust.net$blockGenes[[i]]], "Module colors", 
    dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05)
        }
    dev.off()
}
```

```{r mod_eig_cor, eval = FALSE}
u_mod <- colnames(clust.net$MEs)
all.eig.cor <- vector(mode = "list", length = length(u_mod))
names(all.eig.cor) <- u_mod
for(i in 1:length(u_mod)){
  mod.locale <- which(wgcna.modules == gsub("ME", "", u_mod[i]))
  mod.expr <- adj.expr[,mod.locale]
  mod.eig <- clust.net$MEs[,i,drop=FALSE]
  #df <- data.frame(mod.eig)
  #pheatmap(mod.expr, show_rownames = FALSE, show_colnames = FALSE, annotation_row = df)
  #pheatmap(cor(mod.expr), show_rownames = FALSE, show_colnames = FALSE)
  eig.cor <- apply(mod.expr, 2, function(x) cor(x, mod.eig))
  #barplot(sort(eig.cor, decreasing = TRUE))
  all.eig.cor[[i]] <- eig.cor
}

boxplot(lapply(all.eig.cor, abs), las = 2)
abline(h = 0)
```

## Module Size

The following plots show the number of genes in each module.

```{r mod_size, fig.height = 10, fig.width = 7}
count_mod_genes <- function(net.obj){
  if(clust.type == "CoExpNet"){
    moduleV <- net.obj$moduleColors
  }
  if(clust.type == "WGCNA"){
    moduleV <- net.obj$colors
  }

  u_mod <- unique(moduleV)
  gene_count <- sapply(u_mod, function(x) length(which(moduleV == x)))
  return(gene_count)
}

mod_sizes <- count_mod_genes(clust.net)
size_order <- order(mod_sizes)

par(mfrow = c(2,1))
barplot(mod_sizes[size_order], las = 2, col = names(mod_sizes)[size_order], 
ylab = "Module Size (genes)")
barplot(log10(mod_sizes[size_order]), las = 2, col = names(mod_sizes)[size_order],
ylab = "log10 Module Size")
```

## Characterize Modules {.tabset .tabset-fade .tabset-pills}

In each tissue we characterized the modules using gprofiler2. We looked 
for enrichment in all GO domains, as well as KEGG and REACTOME pathways.

The following plot show enriched terms for each set of modules.

```{r get_module_genes}
module_genes <- function(net.obj){
  modules <- gsub("ME", "", colnames(net.obj$MEs))
  if(clust.type == "CoExpNet"){
    moduleV <- net.obj$moduleColors
  }
  if(clust.type == "WGCNA"){
    moduleV <- net.obj$colors
  }
  module.genes <- lapply(modules, function(x) names(moduleV)[which(moduleV == x)])
  names(module.genes) <- modules
  return(module.genes)
}

characterize_module <- function(enrich.table, n.terms = 10, 
pval_thresh = 0.05, max.term.size = NULL, order.by = "p_value",
decreasing = FALSE){

  low.info.words <- c("to", "of", "the", "process", "or", "by", "in", "and")

  if(length(enrich.table) == 0){return("none")}
  
  if(length(enrich.table) > 1){
    enrichments <- enrich.table[[1]]
    if(length(max.term.size) > 0){
      term.locale <- which(enrichments[,"term_size"] <= max.term.size)
    }else{
      term.locale <- 1:nrow(enrichments)
    }
    if(length(pval_thresh) > 0){
      sig.locale <- which(enrichments[,"p_value"] <= pval_thresh)
    }else{
      sig.locale <- 1:nrow(enrichments)
    }
    take.locale <- intersect(term.locale, sig.locale)
    all.terms <- enrichments[take.locale,"term_name"]
    if(!is.null(order.by)){
      ordered.terms <- enrichments[order(enrichments[,order.by], decreasing = decreasing),]
    }else{
      ordered.terms <- enrichments
    }
    all.terms <- ordered.terms[,"term_name"]
    term.words <- unique(unlist(strsplit(all.terms, " ")))
    pruned.words <- setdiff(term.words, low.info.words)
    top.words <- pruned.words[1:n.terms]
    mod.description <- paste(top.words, collapse = "-")
    return(mod.description)
  }
}
```


```{r enrichment}
module.genes <- module_genes(clust.net)
saveRDS(module.genes, file.path(results.dir, "Module_Membership.RDS"))
enrich.file <- file.path(results.dir, "enrich.RDS")
if(!file.exists(enrich.file)){
  enrich <- lapply(module.genes, function(x) gost(x, organism = "mmusculus",
  sources = c("GO", "KEGG", "REACTOME")))
  saveRDS(enrich, enrich.file)
}else{
  enrich <- readRDS(enrich.file)
}
mod.names <- sapply(enrich, 
function(x) characterize_module(x, n.terms = 5, order.by = "p_value", max.term.size = 500))
```

```{r plot_enrich, fig.height = 9, fig.width = 9, results = "asis"}

for(i in 1:length(enrich)){
  cat("### Module", names(enrich)[i], "\n")
  plot.enrichment(enrich[[i]], max.term.size = 1000, plot.label = 
  paste(tissue.type, "Module", names(enrich)[i]), num.terms = 30)
  cat("\n\n")
}
```


## Trait-Module Correlation {.tabset .tabset-fade .tabset-pills}

We looked at the correlations between each module eigengene and each trait.

```{r mod_eig}
mod.eig <- as.matrix(clust.net$MEs)
rownames(mod.eig) <- rownames(adj.expr)
saveRDS(mod.eig, file.path(results.dir, "Module_Eigengenes.RDS"))

#align the phenotype and eigengene matrices
mod.pheno <- get.xz(mod.eig, adj.pheno)
```

```{r align_mat}
cor.mat <- matrix(NA, nrow = ncol(mod.pheno$X), ncol = ncol(mod.pheno$Z))
rownames(cor.mat) <- paste0("ME", colnames(mod.pheno$X))
colnames(cor.mat) <- colnames(mod.pheno$Z)

for(i in 1:ncol(mod.pheno$X)){
  for(j in 1:ncol(mod.pheno$Z)){
    #plot.with.model(mod.pheno$X[,i], mod.pheno$Z[,j], report = "cor.test")
    mod.trait.cor <- cor(mod.pheno$X[,i], mod.pheno$Z[,j])
    cor.mat[i,j] <- mod.trait.cor
  }
}
```

```{r plot_cor, fig.width = 10, fig.height = 6}
pheatmap(t(cor.mat))
write.table(cor.mat, 
  file.path(results.dir, paste0("Module_Trait_Correlation_", tissue.type, ".csv")),
  sep = ",", quote = FALSE)
```

```{r cor_plot, results = "asis", echo = FALSE}
#cat(paste0("![](", cor_fig, "){width=100%}\n"))
```


## Module mapping  {.tabset .tabset-fade .tabset-pills}

Map each module eigengene. Compare these to the mapping of the WGCNA 
modules in Keller et al. 2018. They are pretty grassy,
except for islet module skyblue3, which has a whopping QTL. I think this 
module is probably the same as Plum1 in Keller 2018.

```{r read_geno_data}
#load genotype data
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "dataset.DO.Cube.Adipose")))
if(!data.loaded){
  load(here("Data", "dataset.DO.CUBE.multissue.RData"))
}
```

```{r map_modules, fig.width = 12, fig.height = 8}
scan.file <- file.path(results.dir, "qtl.scan.RDS")
if(!file.exists(scan.file)){
 scans <- scan1(genoprobs, mod.eig, kinship = K)
 saveRDS(scans, scan.file)
}else{
  scans  <- readRDS(scan.file)
}

perm.file <- file.path(results.dir, "qtl.scan.perm.RDS")
if(!file.exists(perm.file)){
 scan.perm <- scan1perm(genoprobs, mod.eig, kinship = K, n_perm = 100)
 saveRDS(scan.perm, perm.file)
}else{
  scan.perm  <- readRDS(perm.file)
}
```

Plot QTL scans thresholded at LOD = 5.

```{r plot_scan, fig.width = 12, fig.height = 6}

mod.dist <- dist(t(scans))
mod.order <- hclust(mod.dist)$order
thresh <- scans[,mod.order]
thresh[which(thresh > 8)] <- 8

#scan_fig <- file.path(results.dir, "Module.Scans.jpg")
#jpeg(scan_fig, width = 12, height = 6, units = "in", res = 300)
par(xpd = TRUE)
multilod.plot(thresh, map = map, lod.thresh = 5, border.lwd = 1, 
row.names = colnames(mod.eig)[mod.order], row.name.shift = -nrow(thresh)*0.011, 
row.text.cex = 0.7, mar = c(2, 10, 2, 0))
par(xpd = FALSE)
#dev.off()
```

```{r plot_individ, eval = FALSE}
qtl_overlap <- function(eqtl.table, mqtl.table, transcript.info, bp.buffer = 5e6){
  eqtl.split.pos <- strsplit(eqtl.table[,"marker.id"], "_")
  eqtl.chr <- sapply(eqtl.split.pos, function(x) x[1])
  eqtl.pos <- as.numeric(sapply(eqtl.split.pos, function(x) x[2]))

  overlapping.eqtl <- vector(mode = "list", length = nrow(mqtl.table))

  for(m in 1:nrow(mqtl.table)){
    mqtl.chr  <- mqtl.table[m,"chr"]
    mqtl.pos <- as.numeric(mqtl.table[m,"pos"])*1e6
    mqtl.min <- mqtl.pos - bp.buffer
    mqtl.max <- mqtl.pos + bp.buffer
    chr.overlap <- which(eqtl.chr == mqtl.chr)
    pos.overlap <- intersect(which(eqtl.pos >= mqtl.min), which(eqtl.pos <= mqtl.max))
    em.overlap <- intersect(chr.overlap, pos.overlap)
    overlap.eqtl <- eqtl.table[em.overlap,c(1:3),drop=FALSE]
    marker.idx <- match(overlap.eqtl[,1], transcript.info[,1])
    info.table <- cbind(overlap.eqtl, transcript.info[marker.idx,c("chr", "start")])
    colnames(info.table) <- c("gene.id", "eQTL.id", "eQTL.LOD", "transcript.Chr", "transcript.Pos")
    overlapping.eqtl[[m]] <- info.table
  }
  return(overlapping.eqtl)
}

mod <- "ME52"
mod <- "ME33"
mod <- "ME49"
mod = "ME40"

plot(scans, lodcol = mod, map = map)
mod.qtl.peaks <- as.matrix(find_peaks(scans[,mod,drop=FALSE], map = map, threshold = 8))

mem.idx <- which(names(module.genes) == gsub("ME", "", mod))
mod.genes <- module.genes[[mem.idx]]
gene.info <- as.matrix(gene.table[match(mod.genes, gene.table[[1]]),])

gene.lod <- tissue.exp$lod$additive
gene.idx <- which(gene.lod[[1]] %in% mod.genes)
eqtl.table <- as.matrix(gene.lod[gene.idx,])

overlap.qtl <- qtl_overlap(eqtl.table, mod.qtl.peaks, gene.info, 5e6)

i = 0
i = i + 1
mod.qtl.peaks[i,]
overlap.qtl[[i]]
```

```{r scan_plot, results = "asis", echo = FALSE}
#cat(paste0("![](", scan_fig, "){width=70%}\n"))
```

Find nominally significant ($p \leq 0.05$) QTL based on permutation testing.
Write QTL to a table called QTL.Significant.Peaks.txt.

```{r sig_qtl}
sig.thresh <- summary(scan.perm, alpha = 0.05)

me.qtl <- vector(mode = "list", length = ncol(mod.eig))
names(me.qtl) <- colnames(mod.eig)
for(i in 1:length(sig.thresh)){
  me.qtl[[i]] <- find_peaks(scans[,i,drop=FALSE], map, threshold = sig.thresh[i], 
  prob = 0.95)
}

me.qtl.table <- Reduce("rbind", me.qtl)
ordered.table <- me.qtl.table[order(me.qtl.table[,2]),]
write.table(ordered.table[,2:ncol(ordered.table)], 
file.path(results.dir, "QTL.Significant.Peaks.txt"), sep = "\t", quote = FALSE, 
row.names = FALSE)
```

```{r heritability_test, eval = FALSE}
local_distal <- function(chr1, pos1, chr2, pos2, local_lim = 1e6){
  if(length(chr1) == 0 || length(chr2) == 0 || length(pos1) == 0 || length(pos2) == 0){return(NA)}
  if(any(is.na(c(chr1, pos1, chr2, pos2)))){return(NA)}
  if(chr1 != chr2){return("distal")}
  pos.dist <- abs(pos1 - pos2)
  if(pos.dist > local_lim){return("distal")}else{return("local")}
}

gene.lod <- as.matrix(tissue.exp$lod$additive)
herit <- est_herit(adj.expr, K[[20]])
eqtl.lod <- lapply(colnames(adj.expr), function(x) as.numeric(gene.lod[which(gene.lod[,1] == x),"lod"]))
max.lod <- sapply(eqtl.lod, function(x) if(length(x) > 0){max(x)}else{0})


transcript.chr <- unlist(sapply(colnames(adj.expr), function(x) gene.table[which(gene.table[[1]] == x),"chr"]))
transcript.pos <- unlist(sapply(colnames(adj.expr), function(x) gene.table[which(gene.table[[1]] == x),"middle"]*1e6))
eqtl.pos <- lapply(colnames(adj.expr), function(x) gene.lod[which(gene.lod[,1] == x),2])
eqtl.max.marker <- sapply(1:length(eqtl.pos), function(x) eqtl.pos[[x]][which.max(eqtl.lod[[x]])])
eqtl.chr <- sapply(eqtl.max.marker, function(x) if(length(x) > 0){strsplit(x, "_")[[1]][1]}else{NA})
eqtl.coord <- sapply(eqtl.max.marker, function(x) if(length(x) > 0){as.numeric(strsplit(x, "_")[[1]][2])}else{NA})

eqtl.class <- sapply(1:ncol(adj.expr), 
  function(x) local_distal(transcript.chr[x], transcript.pos[x], eqtl.chr[x], eqtl.coord[x], 2e6))

num.large.qtl <- sapply(eqtl.lod, length)
num.col <- colors.from.values(num.large.qtl, use.pheatmap.colors = TRUE)
#There is definitely a relationship between heritability and LOD score,
#but there are many exceptions.
pdf("~/Desktop/herit.pdf")
plot(herit, max.lod, xlab = "Heritability", ylab = "Maximum LOD Score", 
  col = as.numeric(as.factor(eqtl.class))+2, pch = 16)
plot(herit[which(eqtl.class == "local")], max.lod[which(eqtl.class == "local")], 
  xlab = "Heritability", ylab = "Maximum LOD Score", pch = 16, main = "Local eQTL is Max",
  ylim = c(0, max(max.lod)))
plot(herit[which(eqtl.class == "distal")], max.lod[which(eqtl.class == "distal")], 
  xlab = "Heritability", ylab = "Maximum LOD Score", pch = 16, main = "Distal eQTL is Max",
  ylim = c(0, max(max.lod)))
plot(herit~as.factor(eqtl.class), ylab = "Heritability", main = "Heritability by Local/Distal",
xlab = "")
plot(max.lod~as.factor(eqtl.class), ylab = "LOD", main = "LOD by Local/Distal", xlab = "")
plot(herit, max.lod, col = num.col,
  xlab = "Heritability", ylab = "Maximum LOD Score", pch = 16, 
  main = "Colored by Number of eQTL with LOD > 6")
plot(herit~as.factor(num.large.qtl), ylab = "Heritability", 
  main = "Heritability by number eQTL LOD > 6", xlab = "")
plot(max.lod~as.factor(num.large.qtl), ylab = "LOD", 
  main = "Max LOD by number eQTL LOD > 6", xlab = "")
dev.off()




```