---
title: "testing pseudotime"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

```{r set_type}
#set testing to FALSE to run full gene expression networks and to TRUE
#to run a subset.
is.interactive = FALSE
#is.interactive = TRUE

args <- commandArgs(trailingOnly=T)
tissue.type <- args[1]

if(is.na(tissue.type)){
  tissue.type = "Islet"
}
```

The purpose of this workflow is to further analyze the expression
modules generated by 1.Cluster_Transcripts.Rmd.


```{r load_code}
library("here")
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "destiny")
load_libraries(all.packages)
```

```{r read_data}
exp.file <- here("Data", "RDS_datasets_tissues", paste0(tissue.type, ".RDS"))
tissue.exp <- readRDS(exp.file)
```

Also collect clinical traits.

```{r pheno}
pheno <- read.csv(here("Data", "DO_clinical_phenotypes.csv"), stringsAsFactors = FALSE, row.names = 1)
#subset to numeric traits
num.pheno <- apply(pheno[,11:30], 2, as.numeric)
rownames(num.pheno) <- rownames(pheno)
tissue.covar <- tissue.exp$covar.matrix
adj.pheno <- adjust(num.pheno, tissue.covar)
gene.table <- tissue.exp$annot.mrna
```

## Filter Expression
Filter the expression matrices to include only genes that have at least
a minimum amount of expression. 

Here we use the raw expression matrix to select transcripts from the rank Z
normalized expression matrix. 

```{r adjust_covar}
min.mean = 10
tissue.trans <- which(colMeans(tissue.exp$data$raw) > min.mean)
tissue.expr <- tissue.exp$data$rz[,tissue.trans]
adj.expr <- adjust(tissue.expr, tissue.covar)
```

## DPT

```{r}
common.ind <- intersect(rownames(adj.expr), rownames(num.pheno))
expr.ind.locale  <- match(common.ind, rownames(adj.expr))
pheno.ind.locale <- match(common.ind, rownames(num.pheno))

pheno.set <- AnnotatedDataFrame(data.frame(num.pheno[pheno.ind.locale,]))
expr.set <- as.ExpressionSet(data.frame(adj.expr[expr.ind.locale,]))
expr.set$phenoData <- AnnotatedDataFrame(data.frame(num.pheno[pheno.ind.locale,]))
dm <- DiffusionMap(expr.set, k = 50)
if(is.interactive){quartz()}
plot(dm)


```